### PR TITLE
usb_dw: USB DMA

### PR DESCRIPTION
**Pull request by andyross**
Time to stop fiddling with this and cut bait.  It works as well as the existing FIFO workaround now, with hardware bugs under the same conditions (one-byte packets).  I can't see anything that breaks in existing usage.  The interrupt ordering mess is documented in code to the extent I can and "works" for all the cases I can find to test (even though it can't in the general case).  Pointers for future work to address shortcomings are documented in the commit message and that giant comment.

### CODE REVIEW

File: drivers/usb/device/Kconfig [hunk 1]
```diff
@@ -21,6 +21,21 @@ config USB_DW_IRQ_PRI
...
 	USB interrupt priority.
 
+config USB_DW_DMA
+	bool "Enable Designware USB Controller Buffer DMA mode"
+	default y
+	help
+	Use buffer DMA instead of FIFO mode for I/O to the controller.
+	This will reduce CPU overhead, but (at least right now)
+	requires a 64 byte per-endpoint buffer be allocated out/into
+	which the transferred data must be copied.  The USB framework
```
Comment (jfischer-phytec-iot) [inline drivers/usb/device/Kconfig]: The same with Kinetis USBFSOTG #542 , I've made the buffers configurable, but I'm not that happy. Maybe we can find better solution later.

**Comment by jfischer-phytec-iot** [comment]
Please change the heading to something like: 
drivers: usb: add DMA for DesignWare usb device driver

**Comment by nashif** [comment]
@andyross I am not getting an ACM device when building samples/subsys/usb/cdc_acm with this patch set, how did you test this?

File: drivers/usb/device/Kconfig [hunk 2]
```diff
@@ -21,6 +21,21 @@ config USB_DW_IRQ_PRI
 	help
 	USB interrupt priority.
 
+config USB_DW_DMA
+	bool "Enable Designware USB Controller Buffer DMA mode"
+	default y
+	help
```
Comment (finikorg) [inline drivers/usb/device/Kconfig]: This patch breaks my USB application (yet to be submitted), IMO this should not be enabled by default

**Review by finikorg — COMMENTED** [review]
We should not enable DMA mode by default

**Review by nashif — CHANGES_REQUESTED** [review]
breaks usb on arduino 101, needs to be rebased probably.

**Comment by nashif** [comment]
outdated, closing. Reopen with updates/rebase if needed.

**Pull request closed without merging**